pkg/scheduler/plugins/coescheduling/podGroupSyncHandler/podGroupSyncHandler.go

```diff
@@ -0,0 +1,312 @@
+func NewPodGroupSyncHandler(client kubernetes.Interface,
+	pgInformer schedinformer.PodGroupInformer,
+	podInformer coreinformer.PodInformer,
+	pgClient schedclientset.Interface, podGroupManager *core.PodGroupManager) *PodGroupSyncHandler {
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: client.CoreV1().Events(v1.NamespaceAll)})
+	ctrl := &PodGroupSyncHandler{
+		eventRecorder: broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "PodGroupSyncHandler"}),
+		pgQueue:       workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PodGroup"),
+	}
+
+	pgInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.pgAdded,
+		UpdateFunc: ctrl.pgUpdated,
+	})
+	pgInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    podGroupManager.OnPodGroupAdd,
+		DeleteFunc: podGroupManager.OnPodDelete,
+	})
+
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    ctrl.podAdded,
+		UpdateFunc: ctrl.podUpdated,
+	})
+	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    podGroupManager.OnPodAdd,
+		DeleteFunc: podGroupManager.OnPodDelete,
+	})
+	ctrl.pgLister = pgInformer.Lister()
+	ctrl.podLister = podInformer.Lister()
+	ctrl.pgListerSynced = pgInformer.Informer().HasSynced
+	ctrl.podListerSynced = podInformer.Informer().HasSynced
+	ctrl.pgClient = pgClient
+	return ctrl
+}
+
+// Run starts listening on channel events
+func (ctrl *PodGroupSyncHandler) Run(workers int, stopCh <-chan struct{}) {
+	defer ctrl.pgQueue.ShutDown()
+
+	klog.InfoS("Starting Pod Group SyncHandler")
+	defer klog.InfoS("Shutting Pod Group SyncHandler")
+
+	if !cache.WaitForCacheSync(stopCh, ctrl.pgListerSynced, ctrl.podListerSynced) {
+		klog.ErrorS(nil, "Cannot sync caches")
+		return
+	}
+	klog.InfoS("Pod Group sync finished")
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+// pgAdded reacts to a PG creation
+func (ctrl *PodGroupSyncHandler) pgAdded(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	pg := obj.(*schedv1alpha1.PodGroup)
+	if pg.Status.Phase == schedv1alpha1.PodGroupFinished || pg.Status.Phase == schedv1alpha1.PodGroupFailed {
+		return
+	}
+	// If startScheduleTime - createTime > 2days, do not enqueue again because pod may have been GCed
+	if pg.Status.Scheduled == pg.Spec.MinMember && pg.Status.Running == 0 &&
+		pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour {
+		return
+	}
+	klog.InfoS("Enqueue podGroup ", "podGroup", key)
+	ctrl.pgQueue.Add(key)
+}
+
+// pgUpdated reacts to a PG update
+func (ctrl *PodGroupSyncHandler) pgUpdated(old, new interface{}) {
+	ctrl.pgAdded(new)
+}
+
+// podAdded reacts to a PG creation
+func (ctrl *PodGroupSyncHandler) podAdded(obj interface{}) {
```

Comment:
if pod apply gang by annotation, shall we create a gang crd to display gang status? if yes, you should consider when to delete this gang crd(may be related with delete from gang cache)

Reply:
yes, we can also create the PodGroup CR